### PR TITLE
haskell.compiler.ghcHEAD: support JS backend

### LINKED ISSUE #208879: haskell.compiler.ghcHEAD: get JS backend to work
`ghcHEAD`  (and the to be released GHC 9.6) have a JS backend that will replace `ghcjs`. We should get it to work in order to be able to get rid of the burdensome `ghcjs` 8.10.7. Ideally we can do this via the already existing `pkgsCross.ghcjs` infrastructure, but emscripten is proving to be a little annoying, so it may not be entirely possible.

Upstream [build instructions](https://gitlab.haskell.org/ghc/ghc/-/wikis/javascript-backend).

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Description of changes

Was able to compile a test jsexe (impurely as the package set doesn't work yet, see the last commit for details)! 

To try it out, build `nix-build -A pkgsCross.ghcjs.buildPackages.haskell.compiler.ghcHEAD` on this branch.

Closes #208879.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ghost — APPROVED** [review]
LGTM if the `systemd` maintainers are okay with the changes to that package (see below).

Built on:
- [x] x86_64-linux
- [ ] powerpc64le-linux (in progress, required hacks)

File: pkgs/os-specific/linux/systemd/default.nix
Comment (sternenseemann) [reply pkgs/os-specific/linux/systemd/default.nix:80]: I guess we can revert that later. In principle this change is entirely unproblematic, as the resulting derivation is the exact same one (as can be discerned from the rebuild count).

**Comment by sternenseemann** [comment]
The gtk3 issue will be fixed by #208945.

**Review by Ericson2314 — APPROVED** [review]
Great to see this!

**Pull request merged**
